### PR TITLE
Improve thread-safety

### PR DESCRIPTION
**Pull request by jkrumbiegel**
- add locks to server and files
- create `borrow_file!` mechanism for simplifying concurrency
- format
- add concurrency test

After adding locks almost to everything, I've simplified the approach so that now, you "borrow" a file (which may optionally create it if you specify that) and while it's borrowed, it is inaccessible to any other task. All the functions mutating `File`s are then expected to be run under this borrowing approach, as making them all threadsafe would be much more difficult.

After this PR, I can use the mechanism to add asynchronous timeouts on top.

### CODE REVIEW

File: src/server.jl [hunk 1]
```diff
@@ -42,10 +42,11 @@ end
 
 struct Server
     workers::Dict{String,File}
-
+    lock::ReentrantLock # should be locked for mutation/lookup of the workers dict, not for evaling on the workers. use worker lock for that
+    workerlocks::Dict{String,ReentrantLock}
```
Comment (MichaelHatherly) [inline src/server.jl:46]: What's the reason the worker lock cannot live directly in the `File` objects?

Comment (jkrumbiegel) [reply src/server.jl:46]: I did that because you need to lock access to a file before it exists. But I'm thinking now that you could in principle pass the pre-locked lock to the `File` constructor as well if we wanted to store it in there.

Comment (MichaelHatherly) [reply src/server.jl:46]: It seems to me better to maintain the locks internally to each file, rather than having to maintain two different dicts.

Comment (jkrumbiegel) [reply src/server.jl:46]: I had it like that first, and changed it during some refactoring. But I didn't try it again after arriving at the current solution. I'll give it another shot.

Comment (jkrumbiegel) [reply src/server.jl:46]: hm the problem is that I don't want to do the file initialization within the server lock, because it can potentially lock it for quite a bit of time. But I'd have to do that, because otherwise I can't signal that I'm locking a not-yet-existing `File`. Unless I add a placeholder `PlaceholderFile` or so that can be placed in the workers dict. But that also seems a bit complicated.

Comment (jkrumbiegel) [reply src/server.jl]: Ok as the file initialization takes a second or so and we are not talking about a high-load server, I put file initialization under the server lock after all, which at least simplifies the double-dict structure.

**Review by MichaelHatherly — APPROVED** [review]

File: src/server.jl [hunk 2]
```diff
@@ -892,7 +892,7 @@ function close!(server::Server, path::String)
     borrow_file!(server, path) do file
         Malt.stop(file.worker)
         lock(server.lock) do
-            delete!(server.workers, path)
+            pop!(server.workers, file.path)
```
Comment (MichaelHatherly) [inline src/server.jl:895]: What's the reason for this change?

Comment (jkrumbiegel) [reply src/server.jl:895]: `pop!` errors if the key is not found, `delete!` doesn't. That hid a bug because `path` isn't necessarily the same as the one used to key the dict.

**Review by MichaelHatherly — APPROVED** [review]
